### PR TITLE
Feat: API write

### PR DESCRIPTION
**Pull request by dBucik**
Ability to write via API call

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -73,3 +73,26 @@ Once you have installed SimpleSAMLphp, installing this module is very simple. Fi
...
+```
+  'apiWriteEnabled' => true,
+  'apiWriteUsername' => 'api_writer,
+  'apiWritePassword' => 'ap1Wr1T3rP@S$,
+```
+#### Example request
+```
+curl --request POST \
+  --url https://proxy.com/proxy/module.php/proxystatistics/writeLoginApi.php \
+  --header 'Authorization: Basic encodedCredentials' \
```
Comment (melanger) [inline README.md:89]: basic auth? 😟

File: www/writeLoginApi.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+$authUsername = $_SERVER['PHP_AUTH_USER'] ?? '';
+$authPass = $_SERVER['PHP_AUTH_PW'] ?? '';
+
+$username = $config->getApiWriteUsername();
+$password = $config->getApiWritePassword();
+
+// If we get here, username was provided. Check password.
+if ($authUsername !== $username || $authPass !== $password) {
```
Comment (melanger) [inline www/writeLoginApi.php:34]:
```suggestion
if ($authUsername !== $username || !password_verify($authPass, $passwordHash)) {
```
Direct comparison of plaintext password is insecure, it is vulnerable to timing attacks and the password is in plaintext in config.

Store password's hash in config (result of [password_hash](https://www.php.net/manual/en/function.password-hash.php)) and then use password_verify.

File: www/writeLoginApi.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    exit;
+}
+$config = Config::getInstance();
+
+if (!$config->isApiWriteEnabled()) {
+    Logger::info(
+        'proxystatistics:writeLoginApi - API write called, but disabled in config. Returning 501 response code'
+    );
+    header('HTTP/1.0 501 Not Implemented');
+    echo '<p>Access denied. You did not enter a password.</p>';
```
Comment (melanger) [inline www/writeLoginApi.php:23]: is this intentional misdirection or it the message incorrect for the situation?

File: www/writeLoginApi.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    Logger::info(
+        'proxystatistics:writeLoginApi - API write called with bad credentials (' . $authUsername . ':' . $authPass . ') returning 401 response code'
+    );
+    header('HTTP/1.0 401 Unauthorized');
+    exit;
+}
+
+$data = json_decode(file_get_contents('php://input'), true);
+
+$userId = $data[DatabaseCommand::API_USER_ID] ?? null;
```
Comment (melanger) [inline www/writeLoginApi.php:44]: If invalid JSON is specified, `$data` contains `false` and the next line will produce an error.
```suggestion
$data = json_decode(file_get_contents('php://input'), true, 5, JSON_THROW_ON_ERROR);

$userId = $data[DatabaseCommand::API_USER_ID] ?? null;
```

File: www/writeLoginApi.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+$serviceIdentifier = $data[DatabaseCommand::API_SERVICE_IDENTIFIER] ?? null;
+$serviceName = $data[DatabaseCommand::API_SERVICE_NAME] ?? null;
+$idpIdentifier = $data[DatabaseCommand::API_IDP_IDENTIFIER] ?? null;
+$idpName = $data[DatabaseCommand::API_IDP_NAME] ?? null;
+
+if (null === $userId || null === $serviceIdentifier || null === $serviceName || null === $idpIdentifier || null === $idpName) {
```
Comment (melanger) [inline www/writeLoginApi.php:50]:
```suggestion
if (!empty(array_diff([DatabaseCommand::API_USER_ID, DatabaseCommand::API_SERVICE_IDENTIFIER, DatabaseCommand::API_SERVICE_NAME, DatabaseCommand::API_IDP_IDENTIFIER, DatabaseCommand::API_IDP_NAME], array_keys($data)))) {
```
The content of the variables is not used anyway.

File: lib/Auth/Process/Statistics.php [hunk 6]
```diff
@@ -19,6 +24,12 @@ public function process(&$request)
     {
         $dateTime = new DateTime();
         $dbCmd = new DatabaseCommand();
-        $dbCmd->insertLogin($request, $dateTime);
+        try {
+            $dbCmd->insertLoginFromFilter($request, $dateTime);
+        } catch (Exception $ex) {
```
Comment (melanger) [inline lib/Auth/Process/Statistics.php:29]: Which exception are you trying to catch, this one? https://github.com/dBucik/proxystatistics-simplesamlphp-module/blob/9c21fbb9ab6030af182bf02dc43b41809c0737c6/lib/DatabaseCommand.php#L396

File: lib/Config.php [hunk 7]
```diff
@@ -44,13 +55,24 @@ private function __construct()
...
         $this->tables = $this->config->getArray('tables', []);
         $this->mode = $this->config->getValueValidate(self::MODE, ['PROXY', 'IDP', 'SP', 'MULTI_IDP'], 'PROXY');
+        $this->apiWriteEnabled = $this->config->getBoolean(self::API_WRITE_ENABLED, false);
+        if ($this->apiWriteEnabled) {
+            $this->apiWriteUsername = $this->config->getString(self::API_WRITE_USERNAME);
+            if (null === $this->apiWriteUsername || empty(trim($this->apiWriteUsername))) {
+                throw new Exception('Username for API write cannot be empty');
+            }
+            $this->apiWritePassword = $this->config->getString(self::API_WRITE_PASSWORD);
+            if (null === $this->apiWritePassword || empty(trim($this->apiWritePassword))) {
```
Comment (melanger) [inline lib/Config.php:65]:
```suggestion
            if (empty(trim($this->apiWriteUsername))) {
                throw new Exception('Username for API write cannot be empty');
            }
            $this->apiWritePassword = $this->config->getString(self::API_WRITE_PASSWORD);
            if (empty(trim($this->apiWritePassword))) {
```
[Configuration::getString()](https://github.com/simplesamlphp/simplesamlphp/blob/da0df94ab05da52ce27ab7c2a360567046bf03b2/lib/SimpleSAML/Configuration.php#L719) never returns `null`, so the `null` check is redundant.

File: lib/DatabaseCommand.php [hunk 8]
```diff
@@ -372,4 +414,41 @@ private function getAggregateGroupBy($ids)
...
+    }
+
+    private function getSpName($request)
+    {
+        $displayName = $request['Destination']['UIInfo']['DisplayName']['en'] ?? '';
+        if (empty($displayName)) {
+            $displayName = $request['Destination']['name']['en'] ?? '';
+        }
+
+        return$displayName;
```
Comment (melanger) [inline lib/DatabaseCommand.php:452]:
```suggestion
        return $displayName;
```

**Review by melanger — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (melanger) [reply README.md:89]: How about OIDC or [JWT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-jwt-bearer)? (aka `client_secret_jwt` or `private_key_jwt` in OIDC)

File: www/writeLoginApi.php (hunk 4 above)
Comment (dBucik) [reply www/writeLoginApi.php]: Modified as suggested.

File: README.md (hunk 1 above)
Comment (dBucik) [reply README.md:89]: Will check with LS Proxy team if they would be able to implement it like that ;)

**Pull request merged**